### PR TITLE
Show checkbox where the favourite icon is now

### LINKED ISSUE #6361: Show checkbox where the favorite star is now
As discussed at the conf:

* show the checkbox for selecting an element where currently the star icon is
* add a little star indicator at the top right of the mime type/thumbnail (as on Android and iOS)
* add the "Favourite" action to the three dots menu

cc @nextcloud/designers 

@danxuliu @ChristophWurst Does one of you want to look into this?

### PR DESCRIPTION
**Pull request by danxuliu**
Fixes #6361

The favourite icon is now just a mark shown on the thumbnail of favourited files (before it was also shown for not favourited files with a different colour). The center of the star is slightly below and to the left of the top right corner of the thumbnail instead of exactly at the corner because, at least for me, it looked better that way ;-) To get the right placement of the thumbnail, the favourite mark, the file text and so on I used a lot of element nesting; I do not know if that could noticeably slow down the rendering on very large file lists or not, but it may be something to keep an eye on ;-)

Marking a file as favourite or unmarking it is done now through the file actions menu (the three-dots menu). Unlike other actions in the file actions menu, the favourite action is a toggle; its text (and icon) change between _Remove from favorites_ and _Add to favorites_ depending on whether the file is already favourited or not (those texts where used instead of the shorter _Unfavorite_ and _Favorite_ due to [the lack of translation contexts](https://github.com/nextcloud/server/issues/6662)). The action was placed in the file actions menu between _Move or copy_ and _Download_ (for no special reason, to be honest).

Regarding the checkbox used to select files it was moved visually to the place where the favourite icon was before. In HTML and CSS, however, it was done by prepending a new column to those tables that allow selections; it was easier (at least, for me :P ) with respect to the CSS (as the CSS for the name column relies a lot in floats and absolute positions (just for the record, it was that way before I touched the code :-P), so accomodating a new element inside that column that sometimes had to be shown and sometimes not was difficult, and in fact for the favourite icon it was done by applying margins and left positions when the table had the CSS class `has-favorites`), and conceptually it also seemed like using an explicit column in the table to provide the selection was the most logical approach. Given that the checkbox is no longer placed on the thumbnail now clicking on the thumbnail opens the file, just like clicking on the file name (as requested in https://github.com/nextcloud/server/issues/6361#issuecomment-330064817).

The aforementioned `has-favorites` class was used before also to align the results table of the search with the main table; as the results table did not include the favourite action but the main table could have it or not the results table looked for the `has-favorites` class and, if it was found, it moved its content to the left as needed. The same approach is used now with the `has-selection` class. Also, the horizontal padding of `#searchresults td` was reduced from 19px to 14px (could this affect other apps?) to align the file names of the results with those of the main table (note that before these changes the space between the thumbnail and the file name in the file list changed slightly depending on whether the table had favourites or not, and that the file names in the results were aligned to the file names in tables with favourites, so the results were not perfectly aligned to tables without favourites; now there is always the same distance between thumbnail and file name, and the results are (or should be :-P ) always perfectly aligned :-) ).

Before, all files:
swapcheckbox-before-allfiles

After moving the icon to the thumbnail, all files:
swapcheckbox-afterthumbnail-allfiles

After moving the checkbox to its own column, all files:
swapcheckbox-aftercolumn-allfiles

Before, recents:
swapcheckbox-before-recents

After, recents:
swapcheckbox-aftercolumn-recents

Before, favorites:
swapcheckbox-before-favorites

After, favorites:
swapcheckbox-aftercolumn-favorites

Before, all files with search:
swapcheckbox-before-search

After moving the icon to the thumbnail, all files with search:
swapcheckbox-afterthumbnail-search

After moving the checkbox to its own column, all files with search:
swapcheckbox-aftercolumn-search

When taking these screenshots I found that search results are shown inconsistently (even on the _stable12_ branch) depending on whether you refresh a section (_All files_, _Recents_...) and then search or you come from a different section and then search (sometimes only the results were shown, sometimes the contents of the folder were shown along with the results, sometimes only the header of the table for the contents of the folder was shown along with the results). @jancborchardt when you have some time could you check that and open an issue detailing what should be the expected behaviour (as you know that better than me ;-) )? Thanks :-)

Back to the checkbox and favourites, on narrow screens keeping the checkbox on the thumbnail could be a better approach than having an exclusive column for the checkbox in order to save valuable horizontal space. However, this would bring back the original problem of selection not being intuitive, specially on touch based devices like mobile phones where it is not natural to hover over the elements; of course the file could be selected just by tapping on the thumbnail, but hovering was what revealed the checkbox and conveyed that it was selectable. Instead of only showing the checkboxes on hover, though, it could be enough to always show the checkboxes to more or less fix the discoverability problem (see the _After moving the icon to the thumbnail, all files_ screenshot above).

I leave that as something to be considered by @nextcloud/designers. In any case, if implemented it will probably be better done on a different pull request. And it will surely have to be done by someone else than me, because I already struggled too much and for too long with the CSS in this pull request and thinking of further changes makes me dizzy :-P

And now I summon @jancborchardt and @skjnldsv for review and, probably, CSS fixes ;-) Thanks!

### CODE REVIEW

File: apps/files/css/files.scss [hunk 1]
```diff
@@ -338,22 +337,23 @@ table td.filename .nametext, .uploadtext, .modified, .column-last>span:first-chi
 /* TODO fix usability bug (accidental file/folder selection) */
 table td.filename .nametext {
 	position: absolute;
```
Comment (skjnldsv) [inline apps/files/css/files.scss]: Damn, too many absolute on the filelist! :(
This is a mess!

**Review by skjnldsv — APPROVED** [review]
Very nice work!
Don't pay attention to my msg on the absolute, not related to this pr, was just a overall comment! :p 

Looks like you don't need help in css after all :D

**Comment by skjnldsv** [comment]
Acceptance failed though! :)
https://drone.nextcloud.com/nextcloud/server/1918/233

**Comment by MariusBluem** [comment]
@jancborchardt What do you think? I don’t like this change design-wise and would change the behavior in the recent file listing instead 🤔

**Comment by danxuliu** [comment]
@skjnldsv
>  Very nice work!

Thanks :-D

>  Don't pay attention to my msg on the absolute, not related to this pr, was just a overall comment! :p

I share your pain :-P

>  Looks like you don't need help in css after all :D

The difference is that some CSS changes that took me hours would have taken you minutes :-P

>  Acceptance failed though! :)
https://drone.nextcloud.com/nextcloud/server/1918/233

Mmmm, it works fine on my local machine. I guess that the problem comes from the animation when the right sidebar is opened; when the test finds the element it has certain position, but when it tries to click on it it has moved and now a different element is at that position.

I suppose that it could be "_fixed_" in the same way as other synchronization problems between the test code and the browser: (automatically) trying again :-P But first I have to be able to replicate it... Could you (or anyone else, I can't do it myself as far as I know) set the review to _Request changes_ to prevent the pull request from being merged? This has to be fixed here instead of in a later pull request; otherwise merging this in its current state would cause the acceptance tests to sporadically fail on other pull requests.

@MariusBluem
>  @jancborchardt What do you think? I don’t like this change design-wise and would change the behavior in the recent file listing instead 🤔

Please note that I have changed this because [it was agreed in a design session at the conference](https://github.com/nextcloud/server/issues/6361) ;-)

**Comment by jancborchardt** [comment]
Nice! Just two things:

- [ ] We should show the checkboxes at a lesser opacity unless hovered. This otherwise is very noisy.
- [ ] Can we make sure that the multiselect is possible in all the Files app views? Or would that be a separate pull request?


@MariusBluem as @danxuliu said, this was the result of a proper design discussion and lots of UX confusion, among other things people not knowing how to multiselect stuff, and people being confused because they wanted to open a file/folder via click on the icon/preview but instead it selected.

**Comment by danxuliu** [comment]
@jancborchardt 
>  We should show the checkboxes at a lesser opacity unless hovered. This otherwise is very noisy.

Ok, I will fix this as well as the acceptance test issue in the next days.

>  Can we make sure that the multiselect is possible in all the Files app views? Or would that be a separate pull request?

I would do that in a separate pull request.

**Comment by skjnldsv** [comment]
Any news?

**Comment by danxuliu** [comment]
@jancborchardt
>  We should show the checkboxes at a lesser opacity unless hovered. This otherwise is very noisy.

Mmm, what opacity should I set for unhovered items? 0.3? 0.5? Other value?

Opacity 0.3:

checkbox-opacity-0 3

Opacity 0.5:

checkbox-opacity-0 5

@skjnldsv
>  Any news?

I have noticed that checkboxes were wrong in the trash bin and fixed them (I will push those changes when I push the rest of them). Regarding the acceptance tests they are almost fixed... but there is still one strange corner case that I am still studying. I hope to finish with that soon ;-)

**Comment by jancborchardt** [comment]
@danxuliu nice! Let's go for .3 (or .4) - .5 is a bit too much. :)

**Comment by danxuliu** [comment]
@jancborchardt
>  Let's go for .3 (or .4) - .5 is a bit too much. :)

 Done; 0.3 then ;-)

I have amended the commits to fix the opacity and the trash bin list, and added a new commit to solve the acceptance test issue. Let's wait for Drone's verdict, but they should work now ;-)

For reference, I copy here too the commit message explaining how the acceptance tests issue was solved:

Firefox and Chrome drivers for Selenium refuse to click on an element if the point to be clicked is covered by a different element, [throwing an UnknownError exception with message "Element is not clickable at point ({x}, {y}). Other element would receive the click: {element}"](https://github.com/SeleniumHQ/selenium/blob/selenium-2.53.0/javascript/firefox-driver/js/syntheticMouse.js#L203). Although in general that would be a legit error (as the user would not be able to click on the element) due to a bad layout, sometimes this can be just a temporal issue caused by an animation, in which case there would be no problem once the animation finished and the elements are all in their final location.

Unfortunately, automatically handling those situations in which the problem is caused by an animation by just retrying a few times if the element to be clicked is covered before giving up would probably cause confusion instead of easing test writing.

The reason is that if the center of the element is covered by another one the Firefox driver for Selenium [tries to click on the corners of the element](https://github.com/SeleniumHQ/selenium/blob/selenium-2.53.0/javascript/firefox-driver/js/utils.js#L748) instead. The problem is that the coordinates used for the click are integer values, but Firefox has sub-pixel accuracy, so sometimes (depending on which corner is not covered and whether the left, top, width or height properties of the element to be clicked have a decimal component or not) the clicks silently land on a different HTML element (and that is with squared borders; with round borders they always land on a different HTML element. [That was partially addressed for Selenium 3.0 by clicking first on the edges](https://github.com/SeleniumHQ/selenium/commit/23a24f5d98b2aa0e4537437bdeda6a6bf5d3b527), but it would still fail if the middle point of the edges is covered but not the corners).

It is not possible to fix or even detect all that from the tests (except maybe with some extreme hacks involving accessing private PHP members from Mink and bypassing or replacing the standard JavaScript executed by the Firefox driver with a custom implementation...), so it is not possible to ensure that clicks during an animation will land on the right element (in fact it is not possible even on static elements, although except when the layout is wrong there should be no problem); sometimes retrying a click when the element is covered would solve the problem, sometimes it would cause a different element to be clicked (and sometimes there would be even no retry, as the first click would have silently landed on a different element than the expected one).

Therefore, a different approach must be used. Instead of trying to automatically handle clicks during animations the tests must be written being aware of the problem and thus waiting somehow for the animations that can cause a problem to end before performing the clicks.

**Comment by danxuliu** [comment]
>  Let's wait for Drone's verdict, but they should work now ;-)

🤦‍ I was so focused on the acceptance tests that I forgot to run the unit tests... I will fix them and push again ;-)

**Comment by danxuliu** [comment]
Unit tests fixed and branch rebased onto current master. Drone failure is not related as far as I know. 

Ready for review again! :-)

File: apps/files/js/tagsplugin.js [hunk 2]
```diff
@@ -84,22 +85,38 @@
...
+					var isFavorite = $file.data('favorite') === true;
+
+					if (isFavorite) {
+						return t('files', 'Remove from favorites');
+					}
+
+					// As it is currently not possible to provide a context for
+					// the i18n strings "Add to favorites" was used instead of
+					// "Favorite" to remove the ambiguity between verb and noun
+					// when it is translated.
```
Comment (MorrisJobke) [inline apps/files/js/tagsplugin.js]: It is: just add a comment before the string with:

```
// TRANSLATORS: ...
```

See https://github.com/nextcloud/server/blob/b4f36d44c43aac0efdc6c70ff8e46473341a9bfe/apps/encryption/templates/altmail.php#L10-L11 for example

**Review by MorrisJobke — APPROVED** [review]
Code looks good and works nicely 👍

File: apps/files/js/tagsplugin.js (hunk 2 above)
Comment (danxuliu) [reply apps/files/js/tagsplugin.js]: I did not know about that, thanks for the info :-)

**Comment by skjnldsv** [comment]
Nice work @danxuliu!! 🎉

**Comment by jancborchardt** [comment]
Yup, great stuff @danxuliu! :) Good to have you on Nextcloud

**Comment by danxuliu** [comment]
Thanks a lot! :-D

**Comment by jancborchardt** [comment]
Ah @danxuliu just one thing, is it possible we show the »Add to favorites« / remove action first in the file actions popover?

**Comment by danxuliu** [comment]
@jancborchardt 
>  is it possible we show the »Add to favorites« / remove action first in the file actions popover?

Sure, here you have: #7012 :-)

**Comment by nickvergessen** [comment]
I like the idea of the checkbox being in front on a single column.
But honestly, I think burrowing the "add to favorites" option in the … menu kills the feature.
Would be nice if we could bring this back to the main view instead so it is immediately visible again.

**Comment by nickvergessen** [comment]
cc @jancborchardt ^

**Comment by jancborchardt** [comment]
We could show both columns on big displays, however it's just too much info at the same time then. The focus on the files itself gets lost.

We dicussed this decision over a long time and came to the conclusion that multiselect should be more present than favoriting since it is done more frequently.

Favoriting for example can also be done in the sidebar, whereas multiselect can only be done in the list. And we do the same with the 3-dot-menu in the Android app.

So yeah, it was a tough decision. But if everything is on the forefront and important, then nothing is really important and all just gets drowned out.

cc @nextcloud/designers
